src/main/java/org/jenkins/tools/test/hook/JenkinsTestHarnessHook.java

```diff
@@ -0,0 +1,158 @@
+                    var bis = new BufferedInputStream(is);
+                    var jis = new JarInputStream(bis)) {
+                var manifest = jis.getManifest();
+                if (manifest == null) {
+                    throw new IllegalArgumentException("Failed to read manifest in " + war);
+                }
+                var version = manifest.getMainAttributes().getValue("Implementation-Version");
+                if (version == null) {
+                    throw new IllegalArgumentException("Failed to read Winstone version from manifest in " + war);
+                }
+                return new VersionNumber(version);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException("Failed to read Winstone version in " + war, e);
+        }
+    }
+
+    /**
+     * Determines the version of Jenkins Test Harness to use depending on the original version.
+     */
+    @NonNull
+    static VersionNumber determineNextVersion(@NonNull VersionNumber version) {
+        var first = true;
+        VersionNumber older = null;
+        for (var validVersion : VALID_VERSIONS.stream().map(VersionNumber::new).collect(Collectors.toList())) {
+            if (first) {
+                if (validVersion.isNewerThan(version)) {
+                    return validVersion;
+                }
+                first = false;
+            }
+            if (validVersion.isOlderThan(version)) {
+                older = validVersion;
+            } else {
+                if (older != null) {
+                    return validVersion;
+                }
+            }
+        }
+        // Keep the version as is. If that happens, it means check() method returned true by mistake.
+        return version;
+    }
+
+    @Override
+    public boolean check(@NonNull BeforeExecutionContext context) {
+        var winstoneVersion = getWinstoneVersion(context.getConfig().getWar());
+        if (winstoneVersion.getDigitAt(0) < 7) {
+            // Don't upgrade anything if winstone version is too old.
+            return false;
+        }
+        try {
+            var existingVersion = getPropertyAsVersion(context, PROPERTY_NAME);
+            // If core uses web fragments, we need a version of jth with web fragments support
+            if (!usesWebFragment(context)) {
+                return existingVersion.isOlderThan(new VersionNumber(VERSION_WITH_WEB_FRAGMENTS));
+            } else {
+                return VALID_VERSIONS.stream().map(VersionNumber::new).anyMatch(existingVersion::equals)
+                        || existingVersion.isOlderThan(new VersionNumber(VERSION_WITH_WEB_FRAGMENTS));
+            }
+        } catch (PomExecutionException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public void action(@NonNull BeforeExecutionContext context) throws PluginCompatibilityTesterException {
+        var version = getPropertyAsVersion(context, PROPERTY_NAME);
```

Comment:
Don't forget to call `super.action()` and/or `super.check()` if needed.

Reply:
Not needed.